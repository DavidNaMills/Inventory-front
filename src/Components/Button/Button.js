import React from 'react';
import classes from './button.module.scss';

const allTypes = {
    base: {
        base: classes.button__base,
        shadow: classes.button__base_shadow,
        border: classes.button__base_border,
        skeleton: classes.button__base_skeleton,
    },

    submit: {
        base: classes.button__submit,
        shadow: classes.button__submit_shadow,
        border: classes.button__submit_border,
        skeleton: classes.button__submit_skeleton,
    },

    clear: {
        base: classes.button__clear,
        shadow: classes.button__clear_shadow,
        border: classes.button__clear_border,
        skeleton: classes.button__clear_skeleton,
    },

    danger: {
        base: classes.button__danger,
        shadow: classes.button__danger_shadow,
        border: classes.button__danger_border,
        skeleton: classes.button__danger_skeleton,
    },

    information: {
        base: classes.button__information,
        shadow: classes.button__information_shadow,
        border: classes.button__information_border,
        skeleton: classes.button__information_skeleton,
    },
};


const Button = ({ config = {}, onClick, label = null, skeleton = false, raised = false, full = false, type = 'base', border = false, bold = false, wide = false, isDisabled = false, Icon = null, noStyle=null }) => {
    const fullStyle = [classes.button];
    let which = type;
    const style = {};

    if (isDisabled) {
        fullStyle.push(classes.button__disabled);
    } else if(noStyle) { fullStyle.push(classes.button_noStyle);
    } else {
        if (allTypes[type]) {
            which = type;
            fullStyle.push(allTypes[type].base);
        } else {
            which = 'base';
            fullStyle.push(allTypes['base'].base);
        }

        if (skeleton) { fullStyle.push(allTypes[which].skeleton); }
        if (raised) { fullStyle.push(allTypes[which].shadow); }
        if (border) { fullStyle.push(allTypes[which].border); }
    }
    if (full) { fullStyle.push(classes.button_full); }
    if (bold) { fullStyle.push(classes.button_bold); }
    
    
    if (wide) {
        style.width = '30%';
    }

    return (
        <button
            {...config}
            onClick={onClick}
            style={style}
            className={fullStyle.join(' ')}
            disabled={isDisabled}
        ><p>{label}</p>{Icon && <Icon />}</button>
    )
}

export default React.memo(Button);