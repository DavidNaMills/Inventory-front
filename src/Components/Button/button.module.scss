@import '../../styles/mixins/buttonMixins.module.scss';
@import '../../styles/variables/colors.module.scss';

.button{
    display: flex;
    align-content: center;
    justify-content: center;
    font-size: 16px;
    border: none;
    padding: 10px;
    padding-left: 20px;
    padding-right: 20px;
    border-radius: 2px;

    p{
        padding-top: 5px;
        padding-right: 10px;
    }

    &__disabled{
        background-color: $disabled-back;
        color: $disabled-color;
    }

    &__base{
        @include buttonStyle($base-back, $base-color);
        &:hover{
            @include onHover($base-shadow, $light-text, $base-back);
        }

        &_shadow{@include buttonShadow($base-shadow);}
        &_border{@include buildBorder($base-color);}
        &_skeleton{@include buildSkeleton($base-back, $base-color); }
    };
    
    &__submit{
        @include buttonStyle($submit-back, $submit-color);
        &:hover{
            @include onHover($submit-shadow, $light-text, $submit-back);
        }

        &_shadow{ @include buttonShadow($submit-shadow); }
        &_border{ @include buildBorder($submit-color); }
        &_skeleton{@include buildSkeleton($submit-back, $submit-color); }
    }
    
    &__clear{
        @include buttonStyle($clear-back, $clear-color);
        &:hover{
            @include onHover($clear-shadow, $clear-color, $clear-back);
        }

        &_shadow{ @include buttonShadow($clear-shadow); }
        &_border{ @include buildBorder($clear-color); }
        &_skeleton{@include buildSkeleton($clear-back, $clear-color); }
    }
    
    &__danger{
        @include buttonStyle($danger-back, $danger-color);
        &:hover{
            @include onHover($danger-shadow, $light-text, $danger-back);
        }
        &_shadow{ @include buttonShadow($danger-shadow); }
        &_border{ @include buildBorder($danger-color); }
        &_skeleton{@include buildSkeleton($danger-back, $danger-color); }
    }
    
    &__information{
        @include buttonStyle($information-back, $information-color);
        &:hover{
            @include onHover($information-shadow, $light-text, $information-back);
        }
        &_shadow{ @include buttonShadow($information-shadow); }
        &_border{ @include buildBorder($information-color); }
        &_skeleton{@include buildSkeleton($information-back, $information-color); }
    }

    &_flat{
        border: none;
        box-shadow: none;
        -webkit-box-shadow: none;
        -moz-box-shadow: none;
    }

    &_full{
        width: 100%;
    }

    &_bold{
        font-weight: bold;
    }

    &_noStyle{
        background-color: transparent;
        border: none;
        box-shadow: none;

        &:hover{
            background-color: rgb(212, 212, 212);
            padding-top: 10px;
            padding-bottom: 10px;
            padding-left: 5px;
            padding-right: 10px;
            border-radius: 100px;
        }
    }
}