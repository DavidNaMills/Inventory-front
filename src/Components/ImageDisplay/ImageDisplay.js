import React from 'react';
import classes from './imagedisplay.module.scss';
import noImage from './noImage.jpg';

const ImageDisplay = (props) =>{
    const{url, alt='', size='medium', blur=false, dispNoImage=true} = props;
    const style = [classes.imageDisplay];

    switch(size){
        case 'small': style.push(classes.imageDisplay__small); break;
        case 'medium': style.push(classes.imageDisplay__medium); break;
        case 'large': style.push(classes.imageDisplay__large); break;
        case 'desktop': style.push(classes.imageDisplay__desktop); break;
        default: style.push(classes.imageDisplay__medium); break;
    }
    
    if(blur){
        style.push(classes.imageDisplay__blur);
    }

    return (
        <img src={url ? url : dispNoImage ? noImage : null} alt={alt} className={style.join(' ')}/>
    )
}

export default React.memo(ImageDisplay);