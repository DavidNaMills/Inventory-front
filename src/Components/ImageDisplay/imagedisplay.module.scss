.imageDisplay{
    padding: '5px';
    margin: '10px';
    margin-bottom: '20px';

    &__blur{
        -webkit-box-shadow: 5px 6px 11px 6px rgba(74,74,74,0.76);
        -moz-box-shadow: 5px 6px 11px 6px rgba(74,74,74,0.76);
        box-shadow: 5px 6px 11px 6px rgba(74,74,74,0.76);
    }

    &__small{
        width: auto;
        max-width: 75px;
    }

    &__medium{
        width: auto;
        max-width: 250px;
    }

    &__large{
        width: auto;
        max-width: 300px;
        height: auto
    }

    &__desktop {
        width: auto;
        max-width: 550px;
        height: auto
    }
}