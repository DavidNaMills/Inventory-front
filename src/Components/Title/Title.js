import React from 'react';
import classes from './title.module.scss';


const Title = (props) => {
    const { size = null, message = '', center = false } = props;
    let element = null;

    switch (size) {
        case 'h1':
            element = <h1 className={[classes.title_h1, center ? classes.title__center : null].join(' ')} >{message}</h1>
            break;
        case 'h2':
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
        case 'h3':
            element = <h3 className={[classes.title_h3, center ? classes.title__center : null].join(' ')}>{message}</h3>
            break;
        case 'h4':
            element = <h4 className={[classes.title_h4, center ? classes.title__center : null].join(' ')}>{message}</h4>
            break;
        case 'h5':
            element = <h5 className={[classes.title_h5, center ? classes.title__center : null].join(' ')}>{message}</h5>
            break;
        case 'h6':
            element = <h6 className={[classes.title_h6, center ? classes.title__center : null].join(' ')}>{message}</h6>
            break;
        default:
            element = <h2 className={[classes.title_h2, center ? classes.title__center : null].join(' ')}>{message}</h2>
            break;
    };

    return (
        <div>
            {element}
        </div>
    )
}

export default React.memo(Title);