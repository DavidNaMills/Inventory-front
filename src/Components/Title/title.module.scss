@import '../../styles/variables/colors.module.scss';
@import '../../styles/fonts/fontSizes.module.scss';
@import '../../styles/mixins/title.module.scss';

.title {
    @include setTitle($h2, $title-color);
    
    &_h1{
        @include setTitle($h1, $title-color);
    };

    &_h2{
        @include setTitle($h2, $title-color);
    };
    
    &_h3{
        @include setTitle($h3, $title-color);
    }

    &_h4{
        @include setTitle($h4, $title-color);
    }

    &_h5{
        @include setTitle($h5, $title-color);
    }

    &_h6{
        @include setTitle($h6, $title-color);
    }

    &__center{
        text-align: center;
    }


}