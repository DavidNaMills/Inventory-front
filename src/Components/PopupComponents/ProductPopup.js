import React from 'react';
import classes from './Popup.module.scss';
import { useSelector } from 'react-redux';
import Title from '../Title/Title';
import ImageDisplay from '../ImageDisplay/ImageDisplay';
import Button from '../Button/Button';
import { isMobileOnly } from 'react-device-detect';
import {useLocale} from '../../Context/LocaleContext';

const tempStyle = {
    backgroundColor: 'rgb(237, 237, 237)',
    padding: '50px',
    border: 'solid black 2px',
    display: 'flex',
    flexDirection: 'column',
}

const bold = {
    fontWeight: 'bold'
}

/**
 * later show more info about how many have been sold:
 * current month
 * in total
 */

const ProductPopup = ({ item, onClose, fields }) => {
    const { types } = useSelector(state => state);
    const locale = useLocale();
    // style={tempStyle} 
    return (
        <div className={classes.popup}>
            <div className={classes.popup__mediumSpacing}>
                <Title message={item.name} size='h2' center />
            </div>
            <div className={classes.popup__image}>
                <ImageDisplay
                    url={item.url ? `/productImages/${item.url}` : null}
                    alt={`${item.name}-${item.code}`}
                    size={isMobileOnly ? 'medium' : 'desktop'}
                    blur
                />
            </div>

            <div className={classes.popup__smallSpacing}>
                <label style={bold} >{locale['prodType']}:</label>
                <p>{types.products[item.type].displayValue}</p>
            </div>

            <div className={classes.popup__smallSpacing}>
                <label style={bold}>{locale['code']}:</label>
                <p>{item.code}</p>
            </div>

            {
                item.description &&
                <div className={classes.popup__smallSpacing}>
                    <label style={bold}>{locale['description']}:</label>
                    <p>{item.description}</p>
                </div>
            }
            {
                fields.map((x, y) =>
                    <div key={y} className={classes.popup__smallSpacing}>
                        <label style={bold}>{locale[x]}:</label>
                        <p key={x}>{item[x]}</p>
                    </div>
                )
            }
            <div className={classes.popup__smallSpacing}>
                <Button onClick={onClose} label={locale['close']} full />
            </div>
        </div>
    )
}

export default ProductPopup;