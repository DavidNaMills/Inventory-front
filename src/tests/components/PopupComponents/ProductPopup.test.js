import React from 'react';
import '../../Containers/containerHelper';
import '../../testHelpers/context';

import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import ProductPopup from '../../../Components/PopupComponents/ProductPopup';
import * as device from 'react-device-detect';
import Title from '../../../Components/Title/Title';
import ImageDisplay from '../../../Components/ImageDisplay/ImageDisplay';
import Button from '../../../Components/Button/Button';

const mockClose = jest.fn();
const item = {
    _id: 'dsa',
    name: 'cups',
    url: 'super url',
    code: '456dsa',
    type: 'def',
    description: 'fdsa',
    qtyInStock: '789',
    sellPrice: '45',
    test1: 'hello',
    test2: 'goodbye'
}
const fields = ['qtyInStock', 'sellPrice'];

describe('<ProductPopup/> component test suite', () => {
    describe('default settings tests', () => {
        let wrapper;
        let div;

        beforeEach(()=>{
            wrapper = shallow(<ProductPopup item={item} onClose={mockClose} fields={fields}/>);
            div = wrapper.find('div');
        })

        it('matches snapshot', () => { 
            const comp = renderer.create(<ProductPopup item={item} onClose={mockClose} fields={fields}/>);
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });
        
        it('exists', () => { 
            expect(wrapper.exists).toBeTruthy();
        });

        it('renders a container <div> with classname "popup"', () => { 
            expect(div.length).toBe(9)
            expect(div.at(0).props()).toHaveProperty('className', 'popup');
        });
        
        it('renders a <span> with classname "popup__mediumSpacing" and <Title/> component', () => { 
            expect(div.at(1).props()).toHaveProperty('className', 'popup__mediumSpacing');
            const ttl = div.at(1).find(Title);
            expect(ttl.length).toBe(1);
            expect(ttl.props()).toHaveProperty('message', item.name);
            expect(ttl.props()).toHaveProperty('size', 'h2');
            expect(ttl.props()).toHaveProperty('center', true);
        });
        
        it('renders a <span> with classname "popup__mediumSpacing" and <ImageDisplay/> component. URL is present', () => { 
            expect(div.at(2).props()).toHaveProperty('className', 'popup__mediumSpacing');
            const img = div.at(2).find(ImageDisplay);
            expect(img.length).toBe(1);
            expect(img.props()).toHaveProperty('url', `/productImages/${item.url}`)
            expect(img.props()).toHaveProperty('size', 'desktop');
            expect(img.props()).toHaveProperty('blur', true);
        });
        
        [
            {
                msg: 'renders a container <div> with classname "popup__smallSpacing" with a <label> and <p> displaying type',
                divPos: 3,
                divCls: 'popup__smallSpacing',
                lblProp:{
                    name:'fontWeight', 
                    value:'bold',
                    txt: 'Type:'
                },
                pProp: 'yellow cups'
            },
            {
                msg: 'renders a container <div> with classname "popup__smallSpacing" with a <label> and <p> displaying code',
                divPos: 4,
                divCls: 'popup__smallSpacing',
                lblProp:{
                    name:'fontWeight', 
                    value:'bold',
                    txt: 'Code:'
                },
                pProp: item.code
            },
            {
                msg: 'renders a container <div> with classname "popup__smallSpacing" with a <label> and <p> if description is present',
                divPos: 5,
                divCls: 'popup__smallSpacing',
                lblProp:{
                    name:'fontWeight', 
                    value:'bold',
                    txt: 'Description:'
                },
                pProp: item.description
            },
            {
                msg: `renders a <div> with classname "popup__smallSpacing" and <label> and <p> for item ${fields[0]} in the fields array`,
                divPos: 6,
                divCls: 'popup__smallSpacing',
                lblProp:{
                    name:'fontWeight', 
                    value:'bold',
                    txt: 'qtyInStock:'
                },
                pProp: item.qtyInStock
            },
            {
                msg: `renders a <div> with classname "popup__smallSpacing" and <label> and <p> for item ${fields[1]} in the fields array`,
                divPos: 7,
                divCls: 'popup__smallSpacing',
                lblProp:{
                    name:'fontWeight', 
                    value:'bold',
                    txt: 'sellPrice:'
                },
                pProp: item.sellPrice
            },
        ].forEach(x=>{
            it(x.msg, () => { 
                    expect(div.at(x.divPos).props()).toHaveProperty('className', x.divCls);
                    const lbl = div.at(x.divPos).find('label');
                    const p = div.at(x.divPos).find('p');
                
                    expect(lbl.length).toBe(1);
                    expect(lbl.props().style).toHaveProperty(x.lblProp.name, x.lblProp.value);
                    expect(lbl.text()).toEqual(x.lblProp.txt);
                    expect(p.length).toBe(1);
                    expect(p.text()).toEqual(x.pProp);
                });
            })

        it('renders a container <div> with classname "popup__smallSpacing" and a <Button/> component', () => { 
            expect(div.at(8).props()).toHaveProperty('className', 'popup__smallSpacing');
            const btn = div.at(8).find(Button);
            expect(btn.length).toBe(1);
            expect(btn.props()).toHaveProperty('onClick', mockClose)
            expect(btn.props()).toHaveProperty('label', 'Close');
            expect(btn.props()).toHaveProperty('full', true);
        });
    });

    describe('with isMobile === true and alternative option tests', () => {
        it('renders the <ImageDisplay/> component with a null URL and size is "medium" if isMobileOnly is true', () => { 
            const orig = device.isMobileOnly;
            device.isMobileOnly = true;
            const tempItem = {
                ...item,
                url: null
            }

            const wrapper = shallow(<ProductPopup item={tempItem} onClose={mockClose} fields={fields}/>);
            const img = wrapper.find(ImageDisplay);
            expect(img.length).toBe(1);
            expect(img.props()).toHaveProperty('url', null);
            expect(img.props()).toHaveProperty('size', 'medium');

            device.isMobileOnly = orig;
        });

        it('doesnt render description if description is not present in the item object', ()=>{
            const tempItem = {
                _id: 'dsa',
                name: 'cups',
                url: 'super url',
                code: '456dsa',
                type: 'def',
                qtyInStock: '789',
                sellPrice: '45'
            };
            const wrapper = shallow(<ProductPopup item={tempItem} onClose={mockClose} fields={fields}/>);
            expect(wrapper.find('div').length).toBe(8);
            expect(wrapper.find('div').at(5).find('p').text()).toEqual(item.qtyInStock);
        });
    });
});