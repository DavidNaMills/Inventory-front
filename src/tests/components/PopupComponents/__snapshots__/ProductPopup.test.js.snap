// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPopup/> component test suite default settings tests matches snapshot 1`] = `
<div
  className="popup"
  style={
    Object {
      "backgroundColor": "rgb(237, 237, 237)",
      "border": "solid black 2px",
      "display": "flex",
      "flexDirection": "column",
      "minWidth": "70%",
      "padding": "50px",
    }
  }
>
  <div
    className="popup__mediumSpacing"
  >
    <div>
      <h2
        className="title_h2 title__center"
      >
        cups
      </h2>
    </div>
  </div>
  <div
    className="popup__mediumSpacing"
  >
    <img
      alt="cups-456dsa"
      className="imageDisplay imageDisplay__desktop imageDisplay__blur"
      src="/productImages/super url"
    />
  </div>
  <div
    className="popup__smallSpacing"
  >
    <label
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Type
      :
    </label>
    <p>
      yellow cups
    </p>
  </div>
  <div
    className="popup__smallSpacing"
  >
    <label
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Code
      :
    </label>
    <p>
      456dsa
    </p>
  </div>
  <div
    className="popup__smallSpacing"
  >
    <label
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Description
      :
    </label>
    <p>
      fdsa
    </p>
  </div>
  <div
    className="popup__smallSpacing"
  >
    <label
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Qty
      :
    </label>
    <p>
      789
    </p>
  </div>
  <div
    className="popup__smallSpacing"
  >
    <label
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Sell Price
      :
    </label>
    <p>
      45
    </p>
  </div>
  <div
    className="popup__smallSpacing"
  >
    <button
      className="button button__base button_full"
      disabled={false}
      onClick={[MockFunction]}
      style={Object {}}
    >
      <p>
        Close
      </p>
    </button>
  </div>
</div>
`;
